File: backend/src/models/Field.ts
import mongoose, { Document, Schema, Model } from "mongoose";

// Interface for static methods
interface IFieldModel extends Model<IField> {
  getStatistics(): Promise<any>;
}

export interface IField extends Document {
  farmId: mongoose.Types.ObjectId;
  fieldName: string;
  soilType: string;
  plantingDate: Date;
  growthStage: string;
  // Devices/sensors array for this specific field
  devices: Array<{
    sensorId: string;
    name: string;
    deviceMacAddress?: string;
    status: "active" | "inactive" | "maintenance";
    registeredAt: Date;
    location?: {
      coordinates: [number, number];
      description?: string;
    };
  }>;
  createdAt: Date;
  updatedAt: Date;

  // Instance methods
  addDevice(deviceData: any): Promise<IField>;
  removeDevice(sensorId: string): Promise<IField>;
  updateGrowthStage(): Promise<IField>;
}

const fieldSchema = new Schema<IField>(
  {
    farmId: {
      type: Schema.Types.ObjectId,
      ref: "Farm",
      required: true,
    },
    fieldName: {
      type: String,
      required: true,
    },
    soilType: {
      type: String,
      required: true,
      enum: ["loamy", "sandy", "clay", "silty"],
    },
    plantingDate: {
      type: Date,
      required: true,
    },
    growthStage: {
      type: String,
      required: true,
      default: "VE", // Initial growth stage
    },
    // Devices/sensors array for this field
    devices: [
      {
        sensorId: {
          type: String,
          required: true,
        },
        name: {
          type: String,
          required: true,
        },
        deviceMacAddress: {
          type: String,
        },
        status: {
          type: String,
          enum: ["active", "inactive", "maintenance"],
          default: "active",
        },
        registeredAt: {
          type: Date,
          default: Date.now,
        },
        location: {
          coordinates: {
            type: [Number],
            default: [121.0244, 14.5995], // Default Manila coordinates
          },
          description: {
            type: String,
            default: "Field sensor location",
          },
        },
      },
    ],
  },
  {
    timestamps: true,
    collection: "fields",
  }
);

// Compound indexes for performance
fieldSchema.index({ farmId: 1 });
fieldSchema.index({ fieldName: 1, farmId: 1 }, { unique: true });
fieldSchema.index({ growthStage: 1 });
fieldSchema.index({ "devices.sensorId": 1 });

// Virtual for days since planting
fieldSchema.virtual("daysSincePlanting").get(function () {
  if (!this.plantingDate) return null;
  const now = new Date();
  const diffTime = Math.abs(now.getTime() - this.plantingDate.getTime());
  return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
});

// Pre-save middleware to auto-update growth stage based on planting date
fieldSchema.pre("save", function (next) {
  if (this.plantingDate) {
    const now = new Date();
    const diffTime = Math.abs(now.getTime() - this.plantingDate.getTime());
    const daysSincePlanting = Math.ceil(diffTime / (1000 * 60 * 60 * 24));

    if (daysSincePlanting <= 7)
      this.growthStage = "VE"; // Emergence (0-7 days)
    else if (daysSincePlanting <= 21)
      this.growthStage = "V3"; // 3rd leaf (8-21 days)
    else if (daysSincePlanting <= 42)
      this.growthStage = "V8"; // 8th leaf (22-42 days)
    else if (daysSincePlanting <= 65)
      this.growthStage = "VT"; // Tasseling (43-65 days)
    else if (daysSincePlanting <= 85)
      this.growthStage = "R1"; // Silking (66-85 days)
    else this.growthStage = "R6"; // Maturity (86+ days)
  }
  next();
});

// Static method to get field statistics
fieldSchema.statics.getStatistics = async function () {
  const stats = await this.aggregate([
    {
      $group: {
        _id: null,
        totalFields: { $sum: 1 },
        fieldsWithDevices: {
          $sum: { $cond: [{ $gt: [{ $size: "$devices" }, 0] }, 1, 0] },
        },
        fieldsByGrowthStage: {
          $push: "$growthStage",
        },
        fieldsBySoilType: {
          $push: "$soilType",
        },
      },
    },
  ]);

  return (
    stats[0] || {
      totalFields: 0,
      fieldsWithDevices: 0,
      fieldsByGrowthStage: [],
      fieldsBySoilType: [],
    }
  );
};

// Instance method to add device
fieldSchema.methods.addDevice = function (deviceData: any) {
  this.devices.push({
    sensorId: deviceData.sensorId,
    name: deviceData.name,
    deviceMacAddress: deviceData.deviceMacAddress,
    status: deviceData.status || "active",
    registeredAt: new Date(),
    location: deviceData.location || {
      coordinates: [121.0244, 14.5995],
      description: "Field sensor location",
    },
  });
  return this.save();
};

// Instance method to remove device
fieldSchema.methods.removeDevice = function (sensorId: string) {
  this.devices = this.devices.filter(
    (device: any) => device.sensorId !== sensorId
  );
  return this.save();
};

// Instance method to update growth stage
fieldSchema.methods.updateGrowthStage = function () {
  // Growth stage is automatically updated in pre-save middleware
  return this.save();
};

const Field = mongoose.model<IField, IFieldModel>("Field", fieldSchema);

export default Field;

File: frontend/mobile/lib/features/farm/domain/entities/field.dart
import 'package:equatable/equatable.dart';

class Field extends Equatable {
  final String? id;
  final String farmId;
  final String fieldName;
  final String soilType;
  final DateTime plantingDate;
  final String growthStage;
  final List<Map<String, dynamic>>? devices;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Field({
    this.id,
    required this.farmId,
    required this.fieldName,
    required this.soilType,
    required this.plantingDate,
    required this.growthStage,
    this.devices,
    required this.createdAt,
    required this.updatedAt,
  });

  // Calculate days since planting
  int get daysSincePlanting {
    final now = DateTime.now();
    return now.difference(plantingDate).inDays;
  }

  // Auto-calculate growth stage based on planting date
  String get calculatedGrowthStage {
    final days = daysSincePlanting;
    if (days <= 7) return 'VE'; // Emergence (0-7 days)
    if (days <= 21) return 'V3'; // 3rd leaf (8-21 days)
    if (days <= 42) return 'V8'; // 8th leaf (22-42 days)
    if (days <= 65) return 'VT'; // Tasseling (43-65 days)
    if (days <= 85) return 'R1'; // Silking (66-85 days)
    return 'R6'; // Maturity (86+ days)
  }

  // Get growth stage description
  String get growthStageDescription {
    switch (growthStage) {
      case 'VE':
        return 'Emergence Stage';
      case 'V3':
        return 'Third Leaf Stage';
      case 'V8':
        return 'Eighth Leaf Stage';
      case 'VT':
        return 'Tasseling Stage';
      case 'R1':
        return 'Silking Stage';
      case 'R6':
        return 'Maturity Stage';
      default:
        return 'Unknown Stage';
    }
  }

  // Check if field has devices
  bool get hasDevices => devices != null && devices!.isNotEmpty;

  Field copyWith({
    String? id,
    String? farmId,
    String? fieldName,
    String? soilType,
    DateTime? plantingDate,
    String? growthStage,
    List<Map<String, dynamic>>? devices,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Field(
      id: id ?? this.id,
      farmId: farmId ?? this.farmId,
      fieldName: fieldName ?? this.fieldName,
      soilType: soilType ?? this.soilType,
      plantingDate: plantingDate ?? this.plantingDate,
      growthStage: growthStage ?? this.growthStage,
      devices: devices ?? this.devices,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'farmId': farmId,
      'fieldName': fieldName,
      'soilType': soilType,
      'plantingDate': plantingDate.toIso8601String(),
      'growthStage': growthStage,
      'devices': devices,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  factory Field.fromJson(Map<String, dynamic> json) {
    // Handle the actual backend response structure
    // Backend returns 'sensors' but we expect 'devices'
    List<Map<String, dynamic>>? devices;
    if (json['sensors'] != null) {
      devices = List<Map<String, dynamic>>.from(json['sensors']);
    } else if (json['devices'] != null) {
      devices = List<Map<String, dynamic>>.from(json['devices']);
    }

    // Get soilType from the first sensor if available, otherwise use default
    String soilType = 'loamy'; // default
    if (devices != null && devices.isNotEmpty) {
      soilType = devices.first['soilType'] ?? 'loamy';
    }

    return Field(
      id: json['id'] ?? json['_id'],
      farmId: json['farmId'] ?? '', // This will be set by the parent Farm
      fieldName: json['fieldName'] ?? '',
      soilType: soilType,
      plantingDate: DateTime.parse(json['plantingDate']),
      growthStage: json['growthStage'] ?? 'VE',
      devices: devices,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : DateTime.now(),
    );
  }

  @override
  List<Object?> get props => [
    id,
    farmId,
    fieldName,
    soilType,
    plantingDate,
    growthStage,
    devices,
    createdAt,
    updatedAt,
  ];

  @override
  String toString() {
    return 'Field('
        'id: $id, '
        'farmId: $farmId, '
        'fieldName: $fieldName, '
        'soilType: $soilType, '
        'plantingDate: $plantingDate, '
        'growthStage: $growthStage, '
        'devices: ${devices?.length ?? 0} devices, '
        'createdAt: $createdAt, '
        'updatedAt: $updatedAt'
        ')';
  }
}
